File: Ogury/SASOguryBannerAdapter.m
//
//  SASOguryBannerAdapter.m
//  AdViewer
//
//  Created by Loïc GIRON DIT METAZ on 01/10/2020.
//  Copyright © 2020 Smart AdServer. All rights reserved.
//

#import "SASOguryBannerAdapter.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SASOguryBannerAdapter

- (nonnull instancetype)initWithDelegate:(nonnull id<SASMediationBannerAdapterDelegate>)delegate {
    if (self = [super init]) {
        self.delegate = delegate;
    }
    return self;
}

- (void)requestBannerWithServerParameterString:(nonnull NSString *)serverParameterString clientParameters:(nonnull NSDictionary *)clientParameters viewController:(nonnull UIViewController *)viewController {
    // Ogury configuration
    [self configureOgurySDKWithServerParameterString:serverParameterString andClientParameters:clientParameters completionHandler:^(NSError *error) {
        if (error == nil) {
            // Checking the bannerSize parameter
            if (self.bannerSize == nil) {
                // Configuration can fail if the serverParameterString does not contains a valid banner size
                NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeInvalidParameterString userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Invalid server parameter string: %@", serverParameterString] }];
                [self.delegate mediationBannerAdapter:self didFailToLoadWithError:error noFill:NO];
                return;
            }
            
            // Banner loading
            self.banner = [[OguryAdsBanner alloc] initWithAdUnitID:self.adUnitId];
            self.banner.bannerDelegate = self;
            
            // Banner resizing to the size of the Smart banner view
            CGSize smartBannerSize = ((NSValue *)[clientParameters objectForKey:SASMediationClientParameterAdViewSize]).CGSizeValue;
            self.banner.frame = CGRectMake(0, 0, smartBannerSize.width, smartBannerSize.height);
            
            [self.banner loadWithSize:self.bannerSize];
        } else {
            // Configuration can fail if the serverParameterString is invalid or if the Ogury SDK does not initialize properly
            [self.delegate mediationBannerAdapter:self didFailToLoadWithError:error noFill:NO];
        }
    }];    
}

#pragma mark - Ogury banner delegate

- (void)oguryAdsBannerAdAvailable:(OguryAdsBanner*)bannerAds { }

- (void)oguryAdsBannerAdNotAvailable:(OguryAdsBanner*)bannerAds {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdNotAvailable userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Banner - Ad not available"] }];
    [self.delegate mediationBannerAdapter:self didFailToLoadWithError:error noFill:YES];
}

- (void)oguryAdsBannerAdLoaded:(OguryAdsBanner*)bannerAds {
    [self.delegate mediationBannerAdapter:self didLoadBanner:self.banner];
}

- (void)oguryAdsBannerAdNotLoaded:(OguryAdsBanner*)bannerAds {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdNotLoaded userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Banner - Ad not loaded"] }];
    [self.delegate mediationBannerAdapter:self didFailToLoadWithError:error noFill:NO];
}

- (void)oguryAdsBannerAdDisplayed:(OguryAdsBanner*)bannerAds { }

- (void)oguryAdsBannerAdClosed:(OguryAdsBanner*)bannerAds { }

- (void)oguryAdsBannerAdClicked:(OguryAdsBanner*)bannerAds {
    [self.delegate mediationBannerAdapterDidReceiveAdClickedEvent:self];
}

- (void)oguryAdsBannerAdError:(OguryAdsErrorType)errorType forBanner:(OguryAdsBanner*)bannerAds {
    NSError *error = [NSError errorWithDomain:SASOguryAdapterErrorDomain code:SASOguryAdapterErrorCodeAdError userInfo:@{ NSLocalizedDescriptionKey: [NSString stringWithFormat:@"Ogury Banner - Ad error with type: %ld", errorType] }];
    [self.delegate mediationBannerAdapter:self didFailToLoadWithError:error noFill:NO];
}

@end

NS_ASSUME_NONNULL_END

File: Ogury/SASOguryBannerAdapter.swift
//
//  SASOguryBannerAdapter.swift
//  SmartAdServer
//
//  Created by Loïc GIRON DIT METAZ on 22/07/2021.
//  Copyright © 2021 Smart AdServer. All rights reserved.
//

import UIKit
import SASDisplayKit
import OguryAds

/**
 Ogury banner adapter class for Smart SDK mediation.
 
 Mediation adapter classes can be used to display an ad using a third party SDK directly from
 an insertion handled by Smart.
 
 To use an adapter class, you simply have to add them to your Xcode project and they will
 be automatically instantiated by the Smart SDK if needed.
 */
@objc(SASOguryBannerAdapter)
class SASOguryBannerAdapter: SASOguryBaseAdapter, SASMediationBannerAdapter {
    
    /// A delegate that this adapter must call to provide information about the ad loading status or events to the Smart Display SDK.
    private weak var delegate: SASMediationBannerAdapterDelegate? = nil
    
    /// The currently loaded Ogury banner if any, nil otherwise.
    private var banner: OguryBannerAd? = nil
    
    /// The view controller currently used to display the banner if any, nil otherwise.
    private var viewController: UIViewController? = nil
    
    required init(delegate: SASMediationBannerAdapterDelegate) {
        self.delegate = delegate
        super.init()
    }
    
    func requestBanner(withServerParameterString serverParameterString: String, clientParameters: [AnyHashable : Any], viewController: UIViewController) {
        
        self.viewController = viewController
        
        // Ogury configuration
        configureOgurySDK(serverParameterString: serverParameterString, clientParameters: clientParameters) { [self] error in
            if let error = error {
                
                // Configuration can fail if the serverParameterString is invalid or if the Ogury SDK does not initialize properly
                delegate?.mediationBannerAdapter(self, didFailToLoadWithError: error, noFill: false)
                
            } else {
                
                // Checking the bannerSize parameter
                guard let bannerSize = bannerSize else {
                    // Configuration can fail if the serverParameterString does not contains a valid banner size
                    let error = NSError(
                        domain: ErrorConstants.errorDomain,
                        code: ErrorConstants.errorCodeInvalidParameterString,
                        userInfo: [NSLocalizedDescriptionKey: "Invalid server parameter string: \(serverParameterString)"]
                    )
                    delegate?.mediationBannerAdapter(self, didFailToLoadWithError: error, noFill: false)
                    return
                }
                
                // Banner loading
                banner = OguryBannerAd(adUnitId: adUnitId!)
                banner?.delegate = self
                
                // Banner resizing to the size of the Smart banner view
                if let smartBannerSize = (clientParameters[SASMediationClientParameterAdViewSize] as? NSValue)?.cgSizeValue {
                    banner?.frame = CGRect(x: 0.0, y: 0.0, width: smartBannerSize.width, height: smartBannerSize.height)
                }
                
                banner?.load(with: bannerSize)
            }
        }
    }
    
}

/**
 Ogury delegate implementation.
 */
extension SASOguryBannerAdapter : OguryBannerAdDelegate {
    
    func didLoad(_ banner: OguryBannerAd) {
        delegate?.mediationBannerAdapter(self, didLoadBanner: banner)
    }
    
    func didDisplay(_ banner: OguryBannerAd) { }
    
    func didClick(_ banner: OguryBannerAd) {
        delegate?.mediationBannerAdapterDidReceiveAdClickedEvent(self)
    }
    
    func didClose(_ banner: OguryBannerAd) { }
    
    func didFailOguryBannerAdWithError(_ error: OguryError, for banner: OguryBannerAd) {
        let error = NSError(
            domain: ErrorConstants.errorDomain,
            code: ErrorConstants.errorCodeAdError,
            userInfo: [NSLocalizedDescriptionKey: "Ogury Banner - Ad failed with error: \(error)"]
        )
        delegate?.mediationBannerAdapter(self, didFailToLoadWithError: error, noFill: (error.code == ErrorConstants.oguryNoAdErrorCode))
    }
    
    func didTriggerImpressionOguryBannerAd(_ banner: OguryBannerAd) { }

    func presentingViewController(forOguryAdsBannerAd banner: OguryBannerAd) -> UIViewController {
        return viewController!;
    }
    
}
